### PR TITLE
Transactionality

### PR DESCRIPTION
**Pull request by danielmarbach**
Currently, we are not yet executing tests against Cosmos on the build server so I ran them against the latest version on this PR. 

image

### CODE REVIEW

**Comment by danielmarbach** [comment]
@SeanFeldman I will address the saga id finder in a dedicated PR. This is good enough for a review and potential merge after the review comments are addressed.

File: src/NServiceBus.Persistence.AzureStorage/Config/CloudTableClientForSubscriptionsFromConnectionString.cs [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+
+    class CloudTableClientForSubscriptionsFromConnectionString : IProvideCloudTableClientForSubscriptions
+    {
+        public CloudTableClientForSubscriptionsFromConnectionString(string subscriptionConnectionString)
+        {
+            var subscriptionAccount = CloudStorageAccount.Parse(subscriptionConnectionString);
+            Client = subscriptionAccount.CreateCloudTableClient();
+            Client.DefaultRequestOptions = new TableRequestOptions
+            {
+                RetryPolicy = new ExponentialRetry()
```
Comment (SeanFeldman) [inline src/NServiceBus.Persistence.AzureStorage/Config/CloudTableClientForSubscriptionsFromConnectionString.cs:13]: There's a discrepancy between Azure databus and ASP. Databus used the `Fixed` retry policy ([link](https://github.com/Particular/NServiceBus.DataBus.AzureBlobStorage/blob/8c2980589a7a929d2a7a3ce1f52544c2c9d3e095/src/DataBus/AzureDataBusPersistence.cs#L52)).

File: src/NServiceBus.Persistence.AzureStorage/Config/IProvideCloudTableClient.cs [hunk 2]
```diff
@@ -0,0 +1,15 @@
+﻿namespace NServiceBus.Persistence.AzureStorage
+{
+    using Microsoft.Azure.Cosmos.Table;
+
+    /// <summary>
+    ///
```
Comment (SeanFeldman) [inline src/NServiceBus.Persistence.AzureStorage/Config/IProvideCloudTableClient.cs:6]: Probably want to fill these out

File: src/NServiceBus.Persistence.AzureStorage/Config/IProvideCloudTableClientForSubscriptions.cs [hunk 3]
```diff
@@ -0,0 +1,15 @@
+﻿namespace NServiceBus.Persistence.AzureStorage
+{
+    using Microsoft.Azure.Cosmos.Table;
+
+    /// <summary>
+    ///
```
Comment (SeanFeldman) [inline src/NServiceBus.Persistence.AzureStorage/Config/IProvideCloudTableClientForSubscriptions.cs:6]: You love these summaries 😜

File: src/NServiceBus.Persistence.AzureStorage/Config/TableEntityPartitionKey.cs [hunk 4]
```diff
@@ -0,0 +1,66 @@
+﻿namespace NServiceBus
+{
+    using System;
+    using Persistence.AzureStorage;
+
+    /// <summary>
+    ///
```
Comment (SeanFeldman) [inline src/NServiceBus.Persistence.AzureStorage/Config/TableEntityPartitionKey.cs:7]: While the name is self-evident, probably a good idea to let the summary describe what this stands for.

File: src/NServiceBus.Persistence.AzureStorage/Config/TableInformation.cs [hunk 5]
```diff
@@ -0,0 +1,26 @@
+﻿namespace NServiceBus
+{
+    using Persistence.AzureStorage;
+
+    /// <summary>
+    ///
```
Comment (SeanFeldman) [inline src/NServiceBus.Persistence.AzureStorage/Config/TableInformation.cs:6]: You've got a TODO, so assuming just didn't get to it.

File: src/NServiceBus.Persistence.AzureStorage/Config/ThrowIfNoCloudTableClientForSubscriptionsProvider.cs [hunk 6]
```diff
@@ -0,0 +1,11 @@
+﻿namespace NServiceBus.Persistence.AzureStorage
+{
+    using System;
+    using Microsoft.Azure.Cosmos.Table;
+
+    class ThrowIfNoCloudTableClientForSubscriptionsProvider : IProvideCloudTableClientForSubscriptions
+    {
+        public CloudTableClient Client => throw new Exception(
+            $"No CloudTableClient has been configured. Either provide a connection string, use `persistence.UseCloudTableClient(client)` or register an implementation of `{nameof(IProvideCloudTableClient)}` in the container.");
```
Comment (SeanFeldman) [inline src/NServiceBus.Persistence.AzureStorage/Config/ThrowIfNoCloudTableClientForSubscriptionsProvider.cs:9]: I wonder if we should change the order in the 'Either provide....' sentence to mention the connection string option the last.

File: src/NServiceBus.Persistence.AzureStorage/Config/ThrowIfNoCloudTableClientProvider.cs [hunk 7]
```diff
@@ -0,0 +1,10 @@
+﻿namespace NServiceBus.Persistence.AzureStorage
+{
+    using System;
+    using Microsoft.Azure.Cosmos.Table;
+
+    class ThrowIfNoCloudTableClientProvider : IProvideCloudTableClient
+    {
+        public CloudTableClient Client => throw new Exception($"No CloudTableClient has been configured. Either provide a connection string, use `persistence.UseCloudTableClient(client)` or register an implementation of `{nameof(IProvideCloudTableClient)}` in the container.");
```
Comment (SeanFeldman) [inline src/NServiceBus.Persistence.AzureStorage/Config/ThrowIfNoCloudTableClientProvider.cs:8]: Similar idea to downplay the connection string option.

File: src/NServiceBus.Persistence.AzureStorage/Outbox/OutboxRecord.cs [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+{
+    using System;
+    using System.Collections.Generic;
+    using Outbox;
+    using Newtonsoft.Json;
+    using Microsoft.Azure.Cosmos.Table;
+
+    class OutboxRecord : TableEntity
+    {
+        [IgnoreProperty]
```
Comment (SeanFeldman) [inline src/NServiceBus.Persistence.AzureStorage/Outbox/OutboxRecord.cs:11]: A comment on why we ignore the Id property (name colliding with Cosmos DB built-in property) would be good to have.

File: src/NServiceBus.Persistence.AzureStorage/Outbox/OutboxRecord.cs [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+        }
+
+        public bool Dispatched { get; set; }
+
+        public string TransportOperations { get; set; }
+
+        public string DispatchedAt { get; set; }
+
+        [IgnoreProperty]
+        public TransportOperation[] Operations { get; set; } = Array.Empty<TransportOperation>();
```
Comment (SeanFeldman) [inline src/NServiceBus.Persistence.AzureStorage/Outbox/OutboxRecord.cs:25]: Potentially describing the trick with ignoring transport operations and writing those as a string on serialization and back when deserializing will be good.

File: src/NServiceBus.Persistence.AzureStorage/SagaPersisters/MigrationSettings.cs [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+    /// </summary>
+    public class MigrationSettings : ExposeSettings
+    {
+        internal MigrationSettings(SettingsHolder settings) : base(settings)
+        {
+        }
+
+        /// <summary>
+        /// By default the persistence operates in a backward compatible mode that tries to find sagas by using the secondary index property.
+        /// Once all sagas have been migrated the lookup can be disabled.
```
Comment (SeanFeldman) [inline src/NServiceBus.Persistence.AzureStorage/SagaPersisters/MigrationSettings.cs:18]: Migrated from what version to what? There were already migrations that involved the secondary index. Perhaps would be worth stating migrating from X to later verions.

File: src/NServiceBus.Persistence.AzureStorage/SagaPersisters/SagaIdGenerator.cs [hunk 11]
```diff
@@ -0,0 +1,44 @@
+﻿namespace NServiceBus.Persistence.AzureStorage
+{
+    using System;
+    using Sagas;
+    using System.Security.Cryptography;
+    using System.Text;
+    using Newtonsoft.Json;
+
+    class SagaIdGenerator : ISagaIdGenerator
```
Comment (SeanFeldman) [inline src/NServiceBus.Persistence.AzureStorage/SagaPersisters/SagaIdGenerator.cs:9]: Do we need to make it public?

File: src/NServiceBus.Persistence.AzureStorage/SagaPersisters/SagaIdGenerator.cs [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+    {
+        public Guid Generate(SagaIdGeneratorContext context)
+        {
+            if (context.CorrelationProperty == SagaCorrelationProperty.None)
+            {
+                // TODO: verify!
```
Comment (SeanFeldman) [inline src/NServiceBus.Persistence.AzureStorage/SagaPersisters/SagaIdGenerator.cs:15]: This has been the case in the past. Is it still the case or no longer?

Comment (SeanFeldman) [reply src/NServiceBus.Persistence.AzureStorage/SagaPersisters/SagaIdGenerator.cs:15]: Custom saga finder means multiple properties. We can't do scan on multiple properties.

File: src/NServiceBus.Persistence.AzureStorage/SynchronizedStorage/IAzureStorageStorageSession.cs [hunk 13]
```diff
@@ -0,0 +1,25 @@
+﻿namespace NServiceBus.Persistence.AzureStorage
+{
+    using Microsoft.Azure.Cosmos.Table;
+
+    /// <summary>
+    ///
```
Comment (SeanFeldman) [inline src/NServiceBus.Persistence.AzureStorage/SynchronizedStorage/IAzureStorageStorageSession.cs:6]: doco :)

File: src/NServiceBus.Persistence.AzureStorage/Testing/TestableAzureStorageStorageSession.cs [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+            var contextBag = new ContextBag();
+            contextBag.Set(partitionKey);
+            ((IWorkWithSharedTransactionalBatch) this).CurrentContextBag = contextBag;
+            PartitionKey = partitionKey.PartitionKey;
+        }
+
+        ContextBag IWorkWithSharedTransactionalBatch.CurrentContextBag { get; set; }
+
+        /// <summary>
+        ///
```
Comment (SeanFeldman) [inline src/NServiceBus.Persistence.AzureStorage/Testing/TestableAzureStorageStorageSession.cs:27]: C'mon Daniel, you love <summary>s 😉

**Review by SeanFeldman — APPROVED** [review]
Approving but be aware that there are a few XML docs items and the build is red.

File: src/NServiceBus.Persistence.AzureStorage/SagaPersisters/SagaIdGenerator.cs (hunk 11 above)
Comment (danielmarbach) [reply src/NServiceBus.Persistence.AzureStorage/SagaPersisters/SagaIdGenerator.cs:9]: I'd say we wait. we can do that if necessary in another PR

File: src/NServiceBus.Persistence.AzureStorage/Config/CloudTableClientForSubscriptionsFromConnectionString.cs (hunk 1 above)
Comment (danielmarbach) [reply src/NServiceBus.Persistence.AzureStorage/Config/CloudTableClientForSubscriptionsFromConnectionString.cs:13]: I know but this is the one we previously had so I left it compliant with what we had before

**Pull request merged**